README.md
Savon
=====

Heavy metal Ruby SOAP client

[Guide](http://rubiii.github.com/savon) | [Rubydoc](http://rubydoc.info/gems/savon) | [Google Group](http://groups.google.com/group/savon-soap) | [Wishlist](http://savon.uservoice.com) | [Bugs](http://github.com/rubiii/savon/issues)

About the eyestreet/savon fork:

This fork is very specific to connecting to the IPAWS OPEN 2.0 server.  All I did was take the carnesmedia/savon fork and modify it to fit my needs.

If you need to do WS-Security, more than likely you will not want to use this fork.  I recommend using the carnesmedia/savon fork and modifying that to fit your needs.

About the carnesmedia/savon branch
----------------------------------

This branch includes a working, but possibly app specific, implementation of WSSESecurity.

To enable the feature, do something like this:

    Savon::Client.new do |wsdl, http, wsse|
      certs = Savon::WSSE::Certs.new :cert_file => "certs/cert_key.pem", :private_key_file => "certs/cert_key.pem", :private_key_password => "a super secret password"
      wsse.sign_with = Savon::WSSE::Signature.new certs
      ...
    end

You'll also need to edit lib/savon/wsse/canonicalizer.rb and add any namespace your documents will have to the <XPath> attributes.

Here's an enumeration of the changes:

### Savon::SOAP::Request

* Process Savon::WSSE::Signature when needed.

### Savon::SOAP::XML

* Changed default env_namespace to soapenv (specific to my application)
* Use WSSE#signature? configuration
* Ability to regenerate (bypassing cached xml)
* Include Savon::WSSE::Signature#body_attributes when signing (through Savon::WSSE)

### Savon::WSSE

* Add #sign_with=/#signature? configuration
* Generate signature xml header when signing
* Include other header attributes even when generating a signature, timestamp, or user_token

### Savon::WSSE::Canonicalizer

* Very *hacky* proxy for xml canonicalization.
* Uses XMLStarlet (http://xmlstar.sourceforge.net/) via the command line, which must be installed seperately.
* Other XMLStarlet alternatives:
 * XMLCanonicalizer (http://rubygems.org/gems/XMLCanonicalizer): DID NOT WORK FOR ME!
 * Nokogiri: For status of c14n support see discussion at https://github.com/tenderlove/nokogiri/issues#issue/226
 * Unofficial fork of libxml-ruby (http://rubygems.org/gems/coupa-libxml-ruby): I haven't tried it yet...

WARNING: This implementation has a bunch of hard-coded xml namespaces, you'll have to add your own.
         Please read the many other comments in lib/savon/wsse/canonicalization.rb.

### Savon::WSSE::Certs
* A class for managing certificates for signing with Savon::WSSE::Signature

### Savon::WSSE::Signature
* Signs soap requests by:
 * adding a wsu:Timestamp
 * digesting the timestamp and the body
 * digitally signing said digests with an X509 certificate
* Requires three passes in order to canonicalize and digest generated sections. (see Savon::SOAP::XML#setup)

### Savon::WSSE::VerifySignature
* Reads a signed soap request/response and attempts to verify the signature
* This class can be used to verify incomming responses, as well as to verify requests generated by Savon::WSSE::Signature for testing.




Installation
------------

Savon is available through [Rubygems](http://rubygems.org/gems/savon) and can be installed via:

    $ gem install savon

Basic workflow
--------------

    # Setting up a Savon::Client representing a SOAP service.
    client = Savon::Client.new do
      wsdl.document = "http://service.example.com?wsdl"
    end

    client.wsdl.soap_actions
    # => [:create_user, :get_user, :get_all_users]

    # Executing a SOAP request to call a "getUser" action.
    response = client.request :get_user do
      soap.body = { :id => 1 }
    end

    response.to_hash
    # => { :get_user_response => { :first_name => "The", :last_name => "Hoff" } }

Excited to learn more?
----------------------

Then you might want to go ahead and read the [Savon Guide](http://rubiii.github.com/savon).

CHANGELOG.md
## 0.8.6 (2011-02-15)

* Fix for issues [issue #147](https://github.com/rubiii/savon/issues/147) and [#151](https://github.com/rubiii/savon/issues/151)
  ([771194](https://github.com/rubiii/savon/commit/771194)).

## 0.8.5 (2011-01-28)

* Fix for [issue #146](https://github.com/rubiii/savon/issues/146) ([98655c](https://github.com/rubiii/savon/commit/98655c)).

* Fix for [issue #147](https://github.com/rubiii/savon/issues/147) ([252670](https://github.com/rubiii/savon/commit/252670)).

## 0.8.4 (2011-01-26)

* Fix for issues [issue #130](https://github.com/rubiii/savon/issues/130) and [#134](https://github.com/rubiii/savon/issues/134)
  ([4f9847](https://github.com/rubiii/savon/commit/4f9847)).

* Fix for [issue #91](https://github.com/rubiii/savon/issues/91) ([5c8ec1](https://github.com/rubiii/savon/commit/5c8ec1)).

* Fix for [issue #135](https://github.com/rubiii/savon/issues/135) ([c9261d](https://github.com/rubiii/savon/commit/c9261d)).

## 0.8.3 (2011-01-11)

* Moved implementation of `Savon::SOAP::Response#to_array` to a class method at `Savon::SOAP::XML.to_array`
  ([05a7d3](https://github.com/rubiii/savon/commit/05a7d3)).

* Fix for [issue #131](https://github.com/rubiii/savon/issues/131) ([4e57b3](https://github.com/rubiii/savon/commit/4e57b3)).

## 0.8.2 (2011-01-04)

* Fix for [issue #127](https://github.com/rubiii/savon/issues/127) ([0eb3da](https://github.com/rubiii/savon/commit/0eb3da4)).

* Changed `Savon::WSSE` to be based on a Hash instead of relying on builder ([4cebc3](https://github.com/rubiii/savon/commit/4cebc3)).

  `Savon::WSSE` now supports wsse:Timestamp headers ([issue #122](https://github.com/rubiii/savon/issues/122)) by setting
  `Savon::WSSE#timestamp` to `true`:

      client.request :some_method do
        wsse.timestamp = true
      end

   or by setting `Savon::WSSE#created_at` or `Savon::WSSE#expires_at`:

      client.request :some_method do
        wsse.created_at = Time.now
        wsse.expires_at = Time.now + 60
      end

  You can also add custom tags to the WSSE header ([issue #69](https://github.com/rubiii/savon/issues/69)):

      client.request :some_method do
        wsse["wsse:Security"]["wsse:UsernameToken"] = { "Organization" => "ACME", "Domain" => "acme.com" }
      end

## 0.8.1 (2010-12-22)

* Update to depend on HTTPI v0.7.5 which comes with a fallback to use Net::HTTP when no other adapter could be required.

* Fix for [issue #72](https://github.com/rubiii/savon/issues/72) ([22074a](https://github.com/rubiii/savon/commit/22074a8)).

* Loosen dependency on builder. Should be quite stable.

## 0.8.0 (2010-12-20)

* Added `Savon::SOAP::XML#env_namespace` ([51fa0e](https://github.com/rubiii/savon/commit/51fa0e)) to configure
  the SOAP envelope namespace. It defaults to :env but can also be set to an empty String for SOAP envelope
  tags without a namespace.

* Replaced quite a lot of core extensions by moving the Hash to XML translation into a new gem called
  [Gyoku](http://rubygems.org/gems/gyoku) ([bac4b4](https://github.com/rubiii/savon/commit/bac4b4)).

## 0.8.0.beta.4 (2010-11-20)

* Fix for [issue #107](https://github.com/rubiii/savon/issues/107) ([1d6eda](https://github.com/rubiii/savon/commit/1d6eda)).

* Fix for [issue #108](https://github.com/rubiii/savon/issues/108)
  ([f64400...0aaca2](https://github.com/rubiii/savon/compare/f64400...0aaca2)) Thanks [fagiani](https://github.com/fagiani).

* Replaced `Savon.response_pattern` with a slightly different implementation of the `Savon::SOAP::Response#to_array` method
  ([6df6a6](https://github.com/rubiii/savon/commit/6df6a6)). The method now accepts multiple arguments representing the response
  Hash keys to traverse and returns the result as an Array or an empty Array in case the key is nil or does not exist.

      response.to_array :get_user_response, :return
      # => [{ :id => 1, :name => "foo"}, { :id => 2, :name => "bar"}]

## 0.8.0.beta.3 (2010-11-06)

* Fix for [savon_spec](http://rubygems.org/gems/savon_spec) to not send nil to `Savon::SOAP::XML#body`
  ([c34b42](https://github.com/rubiii/savon/commit/c34b42)).

## 0.8.0.beta.2 (2010-11-05)

* Added `Savon.response_pattern` ([0a12fb](https://github.com/rubiii/savon/commit/0a12fb)) to automatically walk deeper into
  the SOAP response Hash when a pattern (specified as an Array of Regexps and Symbols) matches the response. If for example
  your response always looks like ".+Response/return" as in:

      <soapenv:
Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
        <soapenv:Body>
          <ns2:authenticateResponse xmlns:ns2="http://v1_0.ws.user.example.com">
            <return>
              <some>thing</some>
            </return>
          </ns2:authenticateResponse>
        </soapenv:Body>
      </soapenv:Envelope>

  you could set the response pattern to:

      Savon.configure do |config|
        config.response_pattern = [/.+_response/, :return]
      end

  then instead of calling:

      response.to_hash[:authenticate_response][:return]  # :some => "thing"

  to get the actual content, Savon::SOAP::Response#to_hash will try to apply given the pattern:

      response.to_hash  # :some => "thing"

  Please notice, that if you don't specify a response pattern or if the pattern doesn't match the
  response, Savon will behave like it always did.

* Added `Savon::SOAP::Response#to_array` (which also uses the response pattern).

## 0.8.0.beta.1 (2010-10-29)

* Changed `Savon::Client.new` to accept a block instead of multiple Hash arguments. You can access the
  wsdl, http and wsse objects inside the block to configure your client for a particular service.

			# Instantiating a client to work with a WSDL document
      client = Savon::Client.new do
        wsdl.document = "http://example.com?wsdl"
      end

			# Directly accessing the SOAP endpoint
			client = Savon::Client.new do
        wsdl.endpoint = "http://example.com"
        wsdl.namespace = "http://v1.example.com"
      end


lib/savon/wsse/signature.rb
require 'rexml/document'
require "savon/wsse/certs"
require 'savon/wsse/canonicalizer'

module Savon
  class WSSE
    class Signature

      class EmptyCanonicalization < RuntimeError; end
      class MissingCertificate < RuntimeError; end

      # For a +Savon::WSSE::Certs+ object. To hold the certs we need to sign.
      attr_accessor :certs

      # Without a document, the document cannot be signed.
      # Generate the document once, and then set document and recall #to_xml
      attr_accessor :document

      ExclusiveXMLCanonicalizationAlgorithm = 'http://www.w3.org/2001/10/xml-exc-c14n#'.freeze
      TransformAlgorithm = 'http://www.w3.org/2001/10/xml-exc-c14n#'.freeze
      RSASHA1SignatureAlgorithm = 'http://www.w3.org/2000/09/xmldsig#rsa-sha1'.freeze
      SHA1DigestAlgorithm = 'http://www.w3.org/2000/09/xmldsig#sha1'.freeze
      RSASHA256SignatureAlgorithm = 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256'.freeze
      SHA256DigestAlgorithm = 'http://www.w3.org/2001/04/xmlenc#sha256'.freeze

      X509v3ValueType = 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509v3'.freeze
      Base64EncodingType = 'http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#Base64Binary'.freeze

      SignatureNamespace = 'http://www.w3.org/2000/09/xmldsig#'.freeze

      def initialize(certs = Certs.new, digest_algorithm = SHA256DigestAlgorithm, signature_algorithm = RSASHA256SignatureAlgorithm)
        @certs = certs
        @digest_algorithm = digest_algorithm
        @signature_algorithm = signature_algorithm
      end

      def have_document?
        !!document
      end

      # Cache "now" so that digests match...
      # TODO: figure out how we might want to expire this cache...
      def now
        @now ||= Time.now
      end

      def timestamp_id
        @timestamp_id ||= "Timestamp-#{uid}".freeze
      end

      def body_id
        @body_id ||= "Body-#{uid}".freeze
      end

      def security_token_id
        @security_token_id ||= "SecurityToken-#{uid}".freeze
      end

      def body_attributes
        {
          "xmlns:wsu" => WSUNamespace,
          "wsu:Id" => body_id,
        }
      end

      def to_xml
        security = {}.deep_merge(timestamp).deep_merge(signature)
        security.deep_merge!(binary_security_token) if certs.cert

        security.merge! :order! => []
        [ "wsse:BinarySecurityToken", "dsig:Signature", "wsu:Timestamp" ].each do |key|
          security[:order!] << key if security[key]
        end

        xml = Gyoku.xml({
          "wsse:Security" => security,
          :attributes! => { "wsse:Security" => {
             'xmlns:wsse' => WSENamespace,
             'xmlns:wsu' => WSUNamespace,
             'soapenv:mustUnderstand' => "1",
          } },
        })
      end

    private

      def binary_security_token
        {
          "wsse:BinarySecurityToken" => Base64.encode64(certs.cert.to_der).gsub("\n", ''),
          :attributes! => { "wsse:BinarySecurityToken" => {
            'ValueType' => X509v3ValueType,
            'EncodingType' => Base64EncodingType,
            "wsu:Id" => security_token_id
          } }
        }
      end

      def signature
        return {} unless have_document?

        sig = signed_info.merge(key_info).merge(signature_value)
        sig.merge! :order! => []
        [ "dsig:SignedInfo", "dsig:SignatureValue", "dsig:KeyInfo" ].each do |key|
          sig[:order!] << key if sig[key]
        end

        {
          "dsig:Signature" => sig,
          :attributes! => { "dsig:Signature" => { 'xmlns:dsig' => SignatureNamespace } },
        }
      end

      def key_info
        {
          "dsig:KeyInfo" => {
            "wsse:SecurityTokenReference" => {
              "wsse:Reference/" => nil,
              :attributes! => { "wsse:Reference/" => {
                "ValueType"  => X509v3ValueType,
                "URI"        => "##{security_token_id}" }
              }
            },
            :attributes! => { "wsse:SecurityTokenReference" => {} },
          },
        }
      end

      def signature_value
        { "dsig:SignatureValue" => the_signature }
      rescue EmptyCanonicalization, MissingCertificate
        {}
      end

      def signed_info
        {
          "dsig:SignedInfo" => {
            "dsig:CanonicalizationMethod/" => nil,
            "dsig:SignatureMethod/" => nil,
            "dsig:Reference" => [
#              signed_info_transforms.merge( signed_info_digest_method ).merge({ "dsig:DigestValue" => timestamp_digest }),
              signed_info_transforms.merge( signed_info_digest_method ).merge({ "dsig:DigestValue" => body_digest }),
            ],
            :attributes! => {
              "dsig:CanonicalizationMethod/" => { "Algorithm" => ExclusiveXMLCanonicalizationAlgorithm },
              "dsig:SignatureMethod/" => { "Algorithm" => @signature_algorithm },
#              "dsig:Reference" => { "URI" => ["##{timestamp_id}", "##{body_id}"] },
              "dsig:Reference" => { "URI" => ["##{body_id}"] },
            },
            :order! => [ "dsig:CanonicalizationMethod/", "dsig:SignatureMethod/", "dsig:Reference" ],
          },
        }
      end

      # We're going to generate the timestamp ourselves, since WSSE is hard-
      # coded to generate the timestamp section directly within wsse:Security.
      #
      # TODO: Allow for configurability of these timestamps.
      def timestamp
        {
          "wsu:Timestamp" => {
            "wsu:Created" => now.xs_datetime,
            "wsu:Expires" => (now + 60 * 5).xs_datetime,
            :order! => ["wsu:Created", "wsu:Expires"],
          },
          :attributes! => { "wsu:Timestamp" => { "wsu:Id" => timestamp_id } },
        }
      end

      def the_signature
        raise MissingCertificate, "Expected a private_key for signing" unless certs.private_key
        xml = canonicalize("SignedInfo")

        case @signature_algorithm
        when RSASHA1SignatureAlgorithm
          signature = certs.private_key.sign(OpenSSL::Digest::SHA1.new, xml)
        when RSASHA256SignatureAlgorithm
          signature = certs.private_key.sign(OpenSSL::Digest::SHA256.new, xml)
        else
          signature = certs.private_key.sign(OpenSSL::Digest::SHA1.new, xml)
        end

        Base64.encode64(signature).gsub("\n", '') # TODO: DRY calls to Base64.encode64(...).gsub("\n", '')
      end

      def timestamp_digest
        xml_digest('wsu:Timestamp')
      end

      def body_digest
        xml_digest("soapenv:Body")
      end

      def canonicalize(xml_element)
        canonicalized_element = Canonicalizer.canonicalize(document, xml_element)
        raise EmptyCanonicalization, "Expected to canonicalize #{xml_element.inspect} within: #{document}" if canonicalized_element.blank?
        canonicalized_element
      end

      def xml_digest(xml_element)
        case @digest_algorithm
        when SHA1DigestAlgorithm
          Base64.encode64(Digest::SHA1.digest( canonicalize( xml_element ))).strip
        when SHA256DigestAlgorithm
          Base64.encode64(Digest::SHA2.digest( canonicalize( xml_element ))).strip
        else
          Base64.encode64(Digest::SHA1.digest( canonicalize( xml_element ))).strip
        end
      end

      def signed_info_digest_method
        { "dsig:DigestMethod/" => nil, :attributes! => { "dsig:DigestMethod/" => { "Algorithm" => @digest_algorithm } } }
      end

      def signed_info_transforms
        { "dsig:Transforms" => { "dsig:Transform/" => nil, :attributes! => { "dsig:Transform/" => { "Algorithm" => ExclusiveXMLCanonicalizationAlgorithm } } } }
      end

      def uid
        OpenSSL::Digest::SHA1.hexdigest([Time.now, rand].collect(&:to_s).join('/'))
      end
    end
  end
end
